### PR TITLE
feat(backend): add federation config properties

### LINKED ISSUE #2375: 🚀 [Feature] [Federation] Allow federation to persist its secret via Config
## 🚀 Feature

Allow federation to persist its secret via Config

- [x] Allow the federation (keypair) to be initialized with a seed
- [x] This seed must be configureable via config
- [x] If the value is not set, select a random seed (does the library do this if you provide `keyPair=null`?).

### PR DESCRIPTION
**Pull request by clauspeterhuebner**
### Issues

- fixes #2375 

### Todo

- [X] None

### CODE REVIEW

File: backend/.env.dist
Comment (ulfgebhardt) [inline backend/.env.dist:74]: is this the secret key or the key seed?

while I agree on `FEDERATE` on the above, this one is part of the DHT hyperswarm right? So it should be `DHT_...`

File: backend/.env.template
Comment (ulfgebhardt) [inline backend/.env.template:65]: you use none of this yet.
Consider to only introduce the secret/seed, nothing else to limit the scope of the PR. Changing the config again is cheap and a well defined process.

If you need the work you done in this PR for followup work, just merge this branch into your next to keep working on that state.

File: backend/.env.dist
Comment (clauspeterhuebner) [reply backend/.env.dist:74]: it is the secret used as seed while creating a  key.

I would prever to rename the DHT_TOPIC to FEDERATE_DHT_TOPIC, because the hyperswarm is the first part of the whole federation process.

File: backend/.env.template
Comment (clauspeterhuebner) [reply backend/.env.template:65]: - Correct in this PR none of them is used,
- the next PR #2339 uses them and for working on, i merged the previous PR in branch of issue #1887
- the third PR #2372 uses then all changes from the previous two PRs and i merged both in the branch of issue #2368 for working on
- the problem will be during working on the 3rd PR to seperate changings belonging to the previous PRs. That means as soon as such uncritical changes can be merged in the master, as easier will be the separation efforts. And the amount of changes can be keept smaller as you all mentioned about my other big PRs ;-)

Comment (ulfgebhardt) [reply backend/.env.template:65]: We will not merge this PR, this is off scope and implements parts which are not used. Please stick to the negotiated scope: I even made a ticket to specify it correctly -> https://github.com/gradido/gradido/issues/2375

File: backend/.env.dist
Comment (ulfgebhardt) [reply backend/.env.dist:74]: That would be ok with me.

But please name the key seed as such, its a better technical description

File: backend/.env.template
Comment (clauspeterhuebner) [reply backend/.env.template:65]: - I followed your requirements to create small PRs with less changed files.
- To go further the next steps will be done on PR #2339 and have to merge the changes from this PR
- To work on the the 3rd step #2372 a merge of all changes from the previous PRs is necessary
- Now my question:
  *  If the previous PRs will not be merged, how can then the amount of changes for the following PRs be kept low?
  * I do not understand what the ticket #2375 will correct against this PR?

Comment (clauspeterhuebner) [reply backend/.env.template:65]: Sorry i recognize, that i wrote this comment, but doesn't save it... ;-)

- Correct in this PR none of them is used,
- the next PR #2339 uses them and for working on, i merged the previous PR in branch of issue #1887
- the third PR #2372 uses then all changes from the previous two PRs and i merged both in the branch of issue #2368 for working on
- the problem will be during working on the 3rd PR to seperate changings belonging to the previous PRs. That means as soon as such uncritical changes can be merged in the master, as easier will be the separation efforts. And the amount of changes can be keept smaller as you all mentioned about my other big PRs ;-)

File: backend/.env.dist
Comment (clauspeterhuebner) [reply backend/.env.dist]: Done

File: backend/.env.dist
Comment (ulfgebhardt) [inline backend/.env.dist:64]:
```suggestion
# Federation
```

File: backend/.env.template
Comment (ulfgebhardt) [inline backend/.env.template:59]:
```suggestion
# Federation
```

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [inline backend/src/federation/index.ts:18]: what we need sodium for?
Just check the config for the required length, dont include sodium to have a definition of `32`.

You can just use `Buffer.alloc(CONFIG.FEDERATE_DHT_SEED, ' hex' )`

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [inline backend/src/federation/index.ts:44]: I think you can omit the function and just write
```suggestion
    const keyPair = DHT.keyPair(Buffer.alloc(CONFIG.FEDERATE_DHT_SEED, ' hex' ))
```

You might require to check length of the secret tho - consider to use the config for this (?) or do the check before

```suggestion
    if(CONFIG.FEDERATE_DHT_SEED.length < DHT.minKeyLength){
      throw new Error()
    }
    const keyPair = DHT.keyPair(...)
```

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [inline backend/src/federation/index.ts:46]: do we want to log the secret key?

File: backend/src/index.ts
Comment (ulfgebhardt) [inline backend/src/index.ts:24]:
```suggestion
    console.log(`Federation active on ${CONFIG.FEDERATE_DHT_TOPIC}`)
```

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [reply backend/src/federation/index.ts:18]: https://github.com/hyperswarm/dht/blob/4190b7505c365ef8a6ad607fc3862780c65eb482/test/connections.js#L554

This one allocs a 32 byte buffer and writes stuff into it - this might be the smartest way?

File: backend/src/federation/index.ts
Comment (clauspeterhuebner) [reply backend/src/federation/index.ts:46]: Not necessary at all.
both keys are logged to see differences between with and without seed creation.
shift the secret-log to debug.level

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [reply backend/src/federation/index.ts:18]: https://github.com/hyperswarm/dht/blob/43b789fd4ed9f12f077350a411bf6cc5c79d6df7/lib/crypto.js#L10

The library does not expose any constant itself. It also just uses plaintext definition of 32 - even tho not ideal I would consider to also do this instead of including sodium

File: backend/.env.dist
Comment (ulfgebhardt) [inline backend/.env.dist:68]: I wondered in your PR already why you call those vars `FEDERATE_` instead of `FEDERATION_`. I think its valid to do so, but I expected the latter there - in the end chose as you please. I think its both valid.

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [reply backend/src/federation/index.ts:18]: if you use `hex` the key must be twice the length: https://www.geeksforgeeks.org/java-program-to-convert-hex-string-to-byte-array/

Two hex chars become 1 byte. 16^2  = 256
You can also choose not to use hex but just plain ascii

Comment (clauspeterhuebner) [reply backend/src/federation/index.ts:18]: i will find a solution without sodium...

File: backend/src/federation/index.ts
Comment (clauspeterhuebner) [reply backend/src/federation/index.ts:44]: found a solution to fix the buffer to the required size of 32 independent from input
i kept the getSeed(), because the Buffer.alloc doesn't handle a set or not set property to return null for the DHT.keyPair invokation.

File: backend/.env.dist
Comment (clauspeterhuebner) [reply backend/.env.dist:68]: renamed to FEDERATION_xxx

File: backend/.env.dist
Comment (ulfgebhardt) [inline backend/.env.dist:1]: do we miss an adjustment in `deployment/bare_metal/.env.dist` ?

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [inline backend/src/federation/index.ts:19]: excluding the log this can be reduced to a one-liner. I agree to have this in a function with the null value as option.

```suggestion
const getSeed = (): Buffer | null => CONFIG.FEDERATION_DHT_SEED ? Buffer.alloc(KEY_SECRET_SEEDBYTES, CONFIG.FEDERATION_DHT_SEED) : null
```

File: backend/src/federation/index.ts
Comment (ulfgebhardt) [inline backend/src/federation/index.ts:21]: funny that this is working - but it must be defined before used - i think that would be more safe and appropriate. Move above the `getSeed()` function.

File: backend/src/index.ts
Comment (ulfgebhardt) [inline backend/src/index.ts:24]: Here you can produce the seed if you think thats important - `starting Federation with seed ...`

File: backend/src/federation/index.ts
Comment (clauspeterhuebner) [reply backend/src/federation/index.ts:19]: good solution, i've tried some expressions with such a one-line logic, but they doesn't work as expected...
now used your one!

File: backend/src/index.ts
Comment (clauspeterhuebner) [reply backend/src/index.ts:24]: This strange behaviour in the keypair with or without seed doesn't appear on using a normal Buffer instead of the sodium malloced one. Don't know why, but with the current solution both key looks quite good or as cryptical as expected.

File: backend/src/federation/index.ts
Comment (clauspeterhuebner) [reply backend/src/federation/index.ts:13]: done

File: backend/src/index.ts
Comment (clauspeterhuebner) [reply backend/src/index.ts]: ...but i add the seed config-info in this log-output

File: backend/.env.dist
Comment (clauspeterhuebner) [reply backend/.env.dist:1]: done

File: deployment/bare_metal/.env.dist
Comment (ulfgebhardt) [inline deployment/bare_metal/.env.dist:65]: I think in the current state of the federation the default should be `disabled`
```suggestion
# FEDERATION_DHT_TOPIC=GRADIDO_HUB
```

**Review by ulfgebhardt — COMMENTED** [review]
Think besides the `default = on` in the deployment config this is ready to merge.

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**
